qiskit/transpiler/passes/synthesis/unitary_synthesis.py

```diff
@@ -0,0 +1,116 @@
+from qiskit.dagcircuit.dagcircuit import DAGCircuit
+from qiskit.circuit.library.standard_gates import iSwapGate, CXGate, CZGate, RXXGate
+from qiskit.extensions.quantum_initializer import isometry
+from qiskit.quantum_info.synthesis.one_qubit_decompose import OneQubitEulerDecomposer
+from qiskit.quantum_info.synthesis.two_qubit_decompose import TwoQubitBasisDecomposer
+
+
+def _choose_kak_gate(basis_gates):
+    """Choose the first available 2q gate to use in the KAK decomposition."""
+
+    kak_gate_names = {
+        'cx': CXGate(),
+        'cz': CZGate(),
+        'iswap': iSwapGate(),
+        'rxx': RXXGate(pi / 2),
+    }
+
+    kak_gate = None
+    kak_gates = set(basis_gates or []).intersection(kak_gate_names.keys())
+    if kak_gates:
+        kak_gate = kak_gate_names[kak_gates.pop()]
+
+    return kak_gate
+
+
+def _choose_euler_basis(basis_gates):
+    """"Choose the first available 1q basis to use in the Euler decomposition."""
+
+    euler_basis_names = {
+        'U3': ['u3'],
+        'U1X': ['u1', 'rx'],
+        'RR': ['r'],
+        'ZYZ': ['rz', 'ry'],
+        'ZXZ': ['rz', 'rx'],
+        'XYX': ['rx', 'ry'],
+    }
+
+    basis_set = set(basis_gates or [])
+
+    for basis, gates in euler_basis_names.items():
+        if set(gates).issubset(basis_set):
+            return basis
+
+    return None
+
+
+class UnitarySynthesis(TransformationPass):
+    """Synthesize gates according to their basis gates."""
+
+    def __init__(self, basis_gates: List[str]):
+        """SynthesizeUnitaries initializer.
+
+        Args:
+            basis_gates: List of gate names to target.
+        """
+        super().__init__()
+        self._basis_gates = basis_gates
+
+    def run(self, dag: DAGCircuit) -> DAGCircuit:
+        """Run the UnitarySynthesis pass on `dag`.
+
+        Args:
+            dag: input dag.
+
+        Returns:
+            Output dag with UnitaryGates synthesized to target basis.
+        """
+
+        euler_basis = _choose_euler_basis(self._basis_gates)
+        kak_gate = _choose_kak_gate(self._basis_gates)
+
+        decomposer1q, decomposer2q = None, None
+        if euler_basis is not None:
+            decomposer1q = OneQubitEulerDecomposer()
+        if kak_gate is not None:
+            decomposer2q = TwoQubitBasisDecomposer(kak_gate, euler_basis=euler_basis)
+
+        for node in dag.named_nodes('unitary'):
+
+            synth_dag = None
+            if len(node.qargs) == 1:
+                if decomposer1q is None:
+                    continue
+                synth_dag = circuit_to_dag(decomposer1q(node.op.to_matrix()))
+            elif len(node.qargs) == 2:
+                if decomposer2q is None:
+                    continue
+                synth_dag = circuit_to_dag(decomposer2q(node.op.to_matrix()))
+            else:
+                synth_dag = circuit_to_dag(
+                    isometry.Isometry(node.op.to_matrix(), 0, 0).definition)
```

Comment:
Instead of having a simple pass where most of the logic is elsewhere, how about we transfer that logic to the pass? This way we can have different passes (methods) for synthesizing an isometry or statevector or unitary or whatever.

Reply:
My thinking was to consolidate the synthesis code under `transpiler/synthesis`, and have the passes that run them (in the context of a DAG) under `transpiler/passes/synthesis` since there are cases where users might want to run synthesis outside of a call to `transpile`.

Reply:
ok I'm fine with that. So will you move it under `transpiler/synthesis` here?